library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java

```diff
@@ -0,0 +1,486 @@
+import org.apache.apex.malhar.lib.window.WindowedOperator;
+import org.apache.apex.malhar.lib.window.WindowedStorage;
+import org.apache.hadoop.classification.InterfaceStability;
+
+import com.google.common.base.Function;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.annotation.InputPortFieldAnnotation;
+import com.datatorrent.common.util.BaseOperator;
+
+/**
+ * This is the abstract windowed operator class that implements most of the windowing, triggering, and accumulating
+ * concepts. The subclass of this abstract class is supposed to provide the implementation of how the accumulated
+ * values are stored in the storage.
+ *
+ */
+@InterfaceStability.Evolving
+public abstract class AbstractWindowedOperator<InputT, OutputT, DataStorageT extends WindowedStorage, AccumulationT extends Accumulation>
+    extends BaseOperator implements WindowedOperator<InputT>
+{
+
+  protected WindowOption windowOption;
+  protected TriggerOption triggerOption;
+  protected long allowedLatenessMillis = -1;
+  protected WindowedStorage<WindowState> windowStateMap;
+
+  private Function<InputT, Long> timestampExtractor;
+
+  private long currentWatermark;
+  private boolean triggerAtWatermark;
+  private long earlyTriggerCount;
+  private long earlyTriggerMillis;
+  private long lateTriggerCount;
+  private long lateTriggerMillis;
+  private long currentDerivedTimestamp = -1;
+  private long windowWidthMillis;
+  protected DataStorageT dataStorage;
+  protected DataStorageT retractionStorage;
+  protected AccumulationT accumulation;
+
+  private static final transient Logger LOG = LoggerFactory.getLogger(AbstractWindowedOperator.class);
+
+  public final transient DefaultInputPort<Tuple<InputT>> input = new DefaultInputPort<Tuple<InputT>>()
+  {
+    @Override
+    public void process(Tuple<InputT> tuple)
+    {
+      processTuple(tuple);
+    }
+  };
+
+  // TODO: This port should be removed when Apex Core has native support for custom control tuples
+  @InputPortFieldAnnotation(optional = true)
+  public final transient DefaultInputPort<ControlTuple> controlInput = new DefaultInputPort<ControlTuple>()
+  {
+    @Override
+    public void process(ControlTuple tuple)
+    {
+      if (tuple instanceof ControlTuple.Watermark) {
+        processWatermark((ControlTuple.Watermark)tuple);
+      }
+    }
+  };
+
+
+  // TODO: multiple input ports for join operations
+
+  public final transient DefaultOutputPort<Tuple<OutputT>> output = new DefaultOutputPort<>();
+
+  // TODO: This port should be removed when Apex Core has native support for custom control tuples
+  public final transient DefaultOutputPort<ControlTuple> controlOutput = new DefaultOutputPort<>();
+
+  /**
+   * Process the incoming data tuple
+   *
+   * @param tuple
+   */
+  public void processTuple(Tuple<InputT> tuple)
+  {
+    long timestamp = extractTimestamp(tuple);
```

Comment:
extractTimestamp method is called here, then in getWindowsValue and also in assignWindow for the same tuple.
If its a TimestampedTuple its fine as as its simple a getter method but if TimeExtractor is set the operation of extracting timestamp might get costly depending on the logic in extractor.

Can this call be made only once?